### PR TITLE
handle copy case of two INTERSECTION type of ATMs

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**

### CODE REVIEW

File: src/checkers/inference/util/CopyUtil.java [hunk 1]
```diff
@@ -154,12 +158,38 @@ private static void copyAnnotationsImpl(final AnnotatedTypeMirror from, final An
...
+            Map<DeclaredType, AnnotatedDeclaredType> toSuperTypesMap = new HashMap<>();
+            for (AnnotatedDeclaredType superType : fromSuperTypes) {
+                fromSuperTypesMap.put(superType.getUnderlyingType(), superType);
+            }
+            for (AnnotatedDeclaredType superType : toSuperTypes) {
+                toSuperTypesMap.put(superType.getUnderlyingType(), superType);
+            }
+
+            //quick check size first
+            if (fromSuperTypesMap.size() != toSuperTypesMap.size()) {
```
Comment (wmdietl) [inline src/checkers/inference/util/CopyUtil.java]: Move this simple check to before creating either of the Maps and check the lengths of fromSuperTypes and toSuperTypes. No point in constructing the Maps if their lengths won't match.

File: src/checkers/inference/util/CopyUtil.java [hunk 2]
```diff
@@ -154,12 +158,38 @@ private static void copyAnnotationsImpl(final AnnotatedTypeMirror from, final An
...
+                ErrorReporter.errorAbort("unequal super types when copying INTERSECTION type! from: " + fromIntersec + " to: " + toIntersec);
+            }
+
+            for (DeclaredType underlyingType : fromSuperTypesMap.keySet()) {
+                if (!toSuperTypesMap.containsKey(underlyingType)) {
+                    ErrorReporter.errorAbort("unequal super types when copying INTERSECTION type! from: " + fromIntersec + " to: " + toIntersec);
+                }
+            }
+
+            //copy annotations in crosponding super types
```
Comment (wmdietl) [inline src/checkers/inference/util/CopyUtil.java]: crosponding -> corresponding

File: src/checkers/inference/util/CopyUtil.java [hunk 3]
```diff
@@ -154,12 +158,38 @@ private static void copyAnnotationsImpl(final AnnotatedTypeMirror from, final An
...
+            for (Entry<DeclaredType, AnnotatedDeclaredType> entry : fromSuperTypesMap.entrySet()) {
+                copyAnnotationsImpl(entry.getValue(), toSuperTypesMap.get(entry.getKey()), copyMethod, visited);
+            }
         } else {
-            // TODO: Hack mode (currently this fails for two INTERSECTION types)
-            if (!InferenceMain.isHackMode()) {
                 ErrorReporter.errorAbort("InferenceUtils.copyAnnotationsImpl: unhandled getKind results: " + from +
```
Comment (wmdietl) [inline src/checkers/inference/util/CopyUtil.java]: Adapt indentation to removed condition.

File: src/checkers/inference/util/CopyUtil.java (hunk 1 above)
Comment (CharlesZ-Chen) [inline src/checkers/inference/util/CopyUtil.java]: Already lift the "size check" up to before constructin maps.

**Comment by wmdietl** [comment]
Thanks!

**Pull request merged**